### ISSUE TITLE
.NET Core - update to RC2

### ISSUE THREAD

**Issue by jeremymeng** [castleproject/Core#165]
after RC2 with dotnet/cli and toolings are released.

 PR is up.

__Update 04/29/2016__

~~Changes for RC2: https://github.com/castleproject/Core/compare/netcore...jeremymeng:rc2~~

- ~~For reference I have a commit for the code changes https://github.com/jeremymeng/Core/commit/96c9c08321d41120e955c7d6235fbf3a53e4ea1d.~~
- The build ~~server~~ script will need some change too.

To-do's:

- [ ] Review API's that are added back to System.Reflection and revert workarounds previously.

**Comment by rjdevereux** [comment]
Is it crazy to assume RC2 will come out during Build?

**Comment by jonorossi** [comment]
Hmm, some of that gets quite a lot more complicated, hopefully `netstandardapp1.5` and 
`netcoreapp1.0` will make more sense later, but [looks](http://stackoverflow.com/questions/35537688/why-theres-both-a-netstandard-and-netstandardapp-tfms) like [others](https://github.com/NuGet/Home/issues/1956) are wondering that now. Don't worry about explaining it now as I'm sure things will change more.

@jeremymeng you shouldn't need to change the build server, this is the script that `build.cmd` runs: https://github.com/castleproject/Core/blob/netcore/buildscripts/Build-NetCore.ps1

**Comment by jeremymeng** [comment]
I moved the dependencies because I thought they would make sense if we are going to target more frameworks in `project.json`.  For switching to netstandard, the actual change should be small.

Ah yes. Only the build script needs an update.

**Comment by jonorossi** [comment]
@jeremymeng ah cool, should we get some of those changes in now?

**Comment by jeremymeng** [comment]
@jonorossi Current VS tooling doesn't support dotnet/cli so the IDE experiences would be broken with these changes.  Maybe wait for a bit until the tooling support is released.

**Comment by jonorossi** [comment]
@jeremymeng I didn't mean the dotnet/cli ones, but the changes to move things around in the project files. If that uses new features, then no problem we can wait.

**Comment by jeremymeng** [comment]
Yes I can get those in first. I just submitted PR #172.

**Comment by pranavkm** [comment]
Maybe tangentially related, the 4.0.0-alpha001 package on nuget.org (https://www.nuget.org/packages/Castle.Core/4.0.0-alpha001) lists the targetFramework as `Unsupported0.0`. This usually happens when you use an old nuget.exe that doesn't understand new TFMs. Just thought I'd let you guys know.

**Comment by jeremymeng** [comment]
Thanks @pranavkm ! It has been fixed by @jonorossi in #174.

**Comment by davidfowl** [comment]
@jeremymeng you should keep the package versions matching the current blessed CoreCLR rc2 build https://github.com/davidfowl/NetStandard/blob/master/corefx.json. I'll be keeping this up to date as we move closer to RC2

**Comment by jeremymeng** [comment]
@davidfowl Thanks for the info.  Is the version number going to be changed again for RC2?

**Comment by kevinkuszyk** [comment]
There's a typo in @davidfowl's link above.  It should be:
https://github.com/davidfowl/NetStandard/blob/master/packages/corefx.json

**Comment by jonorossi** [comment]
@kevinkuszyk not exactly a typo, just GitHub's inability to fix their 404 page to be useful when you move a file.

**Comment by thecodejunkie** [comment]
@jonorossi of topic, but the proper way to get a link from GitHub it to browse to the file you want, and then hit `Y` and it will generate a link to the current blob. That link will never break no matter what, but it will be just that.. a link to that snapshot of the file

In other news.. are there any updates on rc2 compatibility? We have Nancy building / running tests against it right now

**Comment by jeremymeng** [comment]
@thecodejunkie @jonorossi 

I have the rc2 changes ready and it's [building fine](https://ci.appveyor.com/project/jeremymeng/core/build/4.0.0-alpha.32).  I am just waiting for the final RC2 version to settle before sending a PR.

**Comment by pranavkm** [comment]
@jeremymeng the rc2 packages are on MyGet now (https://www.nuget.org/packages/System.Runtime/4.1.0-rc2-24027).

**Comment by jeremymeng** [comment]
@pranavkm Thanks for the notification. I will submit a PR.

@jonorossi Do you know whether the build server support NuGet v3.4.3 or later?  I think we would need that to pack for `.NETStandard1.3`.

**Comment by jonorossi** [comment]
@jeremymeng it was running 3.3.0, just updated TeamCity to use 3.4.3.

**Comment by jeremymeng** [comment]
@jonorossi Any thoughts on a new alpha release that is based on RC2?  NLog and Serilog now have releases that are based on RC2 (except the Serilog.Sinks.ColoredConsole which is still based on RC1).

One benefit is that we will have fixes to packing issue/dependency issues.

**Comment by jonorossi** [comment]
@jeremymeng sorry about not merging the PR, I didn't know you had finished fixing the review comments by your colleague.

I've just merged the PR, I know there were some new APIs added that I logged corefx issues for, it would be great if we could have a run through our conditional compilation symbols and see if things have changed before the next release, especially `FEATURE_NETCORE_REFLECTION_API`.

**Comment by jonorossi** [comment]
@jeremymeng the build fails, `https://raw.githubusercontent.com/dotnet/cli/rel/1.0.0/scripts/obtain/install.ps1` returns 404. Looks like install.ps1 changes to dotnet-install.ps1.

**Comment by pranavkm** [comment]
The file was renamed to `dotnet-install.ps1` (https://github.com/dotnet/cli/blob/rel/1.0.0/scripts/obtain/dotnet-install.ps1). That said, you could just copy the file in to your repo rather than downloading one on the fly.

**Comment by jeremymeng** [comment]
For now I submitted a fix to unblock the build.  I also asked the team about the recommended way for this scenario.  I am not sure if we want to maintain a copy of the script.

**Comment by jeremymeng** [comment]
I update the PR to use a versioned/released version of the script.  In the future the dotnet/cli team will provide a centralized copy of the script that users are encouraged to use.

**Comment by jeremymeng** [comment]
@jonorossi it could be that [the windows dependencies](https://github.com/dotnet/cli/blob/rel/1.0.0/Documentation/cli-prerequisites.md#windows-dependencies) is missing when installing using the script. If installing using the MSI installer these should be chained.  Could you please help update the build server?  I think the dependencies are install by VS 2015 too. Both .NET Core and NET40 passed in [my appveyor build](https://ci.appveyor.com/project/jeremymeng/core/build/4.0.0-alpha.51).

**Comment by jeremymeng** [comment]
@jonorossi there's only one extension method in `FEATURE_NETCORE_REFLECTION_API`, which is to workaround https://github.com/dotnet/corefx/issues/4462. However I took a look and it doesn't seem that the overload you wanted is added. Another issue https://github.com/dotnet/corefx/issues/4463 is about `TargetException`.  It is available now in RC2, but only in `netstandard1.5`.

**Comment by jonorossi** [comment]
> it could be that the windows dependencies is missing when installing using the script

@jeremymeng I installed the VC++ 2012 Runtime, both x86 and x64, but the same error remains. How do we get a verbose output to know where it is looking for this PDB writer binary or package?

**Comment by jonorossi** [comment]
@jeremymeng I found `Microsoft.DiaSymReader.Native.x86.dll` in both:
* `D:\work\d545f799076bb8eb\.dotnet\sdk\1.0.0-preview1-002702\runtimes\win-x86\native`, and
* `C:\Users\teamcity\.nuget\packages\Microsoft.DiaSymReader.Native\1.3.3\runtimes\win-x86\native`

So it looks like it is downloading it fine but the compiler can't find it.

**Comment by jeremymeng** [comment]
@tmat any ideas?  Would using portable PDB workaround the error?  What do we lose if using portable PDB?

**Comment by tmat** [comment]
This has been fixed recently. See https://github.com/dotnet/roslyn/issues/11592. You need Microsoft.DiaSymReader.Native version 1.4.0rc2.

**Comment by jeremymeng** [comment]
@tmat how can we use Microsoft.DiaSymReader.Native version 1.4.0rc2 with dotnet/cli rc2?

**Comment by jeremymeng** [comment]
That sounds too hacky :(.  Maybe we should use portable PDB now while waiting for the next tools release.

**Comment by jeremymeng** [comment]
@jonorossi Installing Visual C++ Redistributable for Visual Studio 2015 seems to fix it.  Is this an acceptable solution?  If not I can submit a PR to use portable PDB.  I have tested both methods on a pristine Windows 7 x86 machine.

**Comment by jonorossi** [comment]
@jeremymeng VCRT 2015 fixed the build thanks. Do we need another `buildOptions` to get RC2 to generate the XML documentation like before?

**Comment by pranavkm** [comment]
Yes - you need to set `xmlDoc: true` (https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Cors/project.json#L18-L21) under `buildOptions`. The compiler will warn \ fail if you have `warningsAsErrors` if you have public types without docs. We add a exclusion for that specific warning in our projects - https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Cors/project.json#L18-L20

**Comment by jeremymeng** [comment]
Yes, I have done some local changes.  There are two additional warnings from xml doc other than CS1591 due to some `cref` type couldn't be resolved in .NET Core build.  I am inclining to rewrite the xml doc comments instead of #if def the `cref`.  Will look at it again on Monday.

**Comment by jeremymeng** [comment]
@jonorossi #181 submitted.

**Comment by jonorossi** [comment]
The PRs have been merged. It doesn't seem like many reflection APIs have been added back yet, so lets just leave it as is for now.

**Comment by jeremymeng** [comment]
Thanks a lot @jonorossi!  Let us know if you are releasing another alpha package.  I can then update downstream libraries so that they can remove "import" : "dotnet5.4", and don't pull in RC1 dependencies any more.

**Comment by jonorossi** [comment]
@jeremymeng will do. I want to make another release but with the logging integration packages this time so Castle Core ships somewhat complete.
